PSA: Phase 3: ECDSA signature verification using PSA

AndrzejKurek opened the pull request:
This PR resolves iotssl-2578 and is created to let other team members review the code.
The helper functions are to be moved to a separate file.
The case of case MBEDTLS_MD_NONE translating to 0 is also to be discussed.
This PR supercedes https://github.com/ARMmbed/mbedtls-psa/pull/174.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    return( PSA_ERROR_INSUFFICIENT_MEMORY );
+}
+
+static psa_algorithm_t translate_md_to_psa( mbedtls_md_type_t md_alg )
```
@mpg @AndrzejKurek Perhaps it's worth creating a separate (header?) file with helper functions like this? I reckon we will need such in other places, too.

AndrzejKurek replied:
Yes, I'd opt for handling it in the target branch.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+#endif
+    case MBEDTLS_MD_NONE:  // Intentional fallthrough
+    default:
+        return( 0 );
```
I am ok with that - it seems that in PSA the value `0` is always represents an invalid slot/algorithm/key.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    size_t len_partial;
+    int tag_type;
+    if( ( end - *p ) < 1 )
+        return( MBEDTLS_ERR_X509_INVALID_SIGNATURE +
```
Minor: Please add curly braces around the `return` statement as it's more than one line long.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    tag_type = **p;
+
+    if( ( ret = mbedtls_asn1_get_tag(p, end, &len_partial,
+    MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 ) {
```
Minor: Please add a line break before `{`.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+
+    if( ( ret = mbedtls_asn1_get_tag( p, end, &len_partial, MBEDTLS_ASN1_INTEGER ) )
+            != 0 )
+        return( ret );
```
The previous branch wraps the error code with `MBEDTLS_ERR_X509_INVALID_SIGNATURE` - should we do the same here?

mpg replied:
Please don't return X509 error codes from the PK module. That would be turning the layers upside down.

AndrzejKurek replied:
Upon encountering an error from this internal function, `MBEDTLS_ERR_PK_BAD_INPUT_DATA` is returned from `ecdsa_verify_wrap`. Should I change this workflow?

hanno-arm replied:
Thanks for pointing that out, @AndrzejKurek. In that case I think we should immediately return unmodified error codes in `extract_ecdsa_sig()` and leave handling them (currently replacing them by `PK_BAD_INPUT_DATA`) to the caller.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+            != 0 )
+        return( ret );
+
+    if( **p == '\0' ) {
```
Minor: Please add a line break before `{`.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+            != 0 )
+        return( ret );
+
+    if( **p == '\0' ) {
```
What's the motivation for this check?

AndrzejKurek replied:
When performing the tests previously, with this function being `mbedtls_x509_get_ecdsa_sig `from https://github.com/AndrzejKurek/mbedtls-psa/pull/2, there were tests where the r or s parts were prefixed with a zero (padding due to DER encoding, if I'm not mistaken?).

hanno-arm replied:
@AndrzejKurek Why would we have such a prefix, and if we have it, why wouldn't we copy it here?

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    }
+
+    sig->p = mbedtls_calloc( 2, len_partial );
+    if( sig->p == NULL ) {
```
Minor: Please add a line break before `{`, or remove the braces altogether.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        MBEDTLS_ERR_ASN1_OUT_OF_DATA );
+    tag_type = **p;
+
+    if( ( ret = mbedtls_asn1_get_tag(p, end, &len_partial,
```
Minor: Space missing after `(`.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    size_t len_signature;
+    size_t len_partial;
+    int tag_type;
+    if( ( end - *p ) < 1 )
```
What's the motivation for this check? All ASN.1 parsing functions include bounds checks. The only thing that IIRC we're not consistently guaranteeing is that they behave correctly if `end > p` (because of checks like `(size_t)( end - * p ) < ... `), but `mbedtls_asn1_get_tag()` does, so I think we should remove the check here.

hanno-arm replied:
@AndrzejKurek I see, it guards the assignment `tag_type = **p` below. However, in contrast to `mbedtls_x509_get_sig()` where the tag is actually stores somewhere, we don't need it here, and it's actually determined as `MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE` through the next call. So I'd suggest to remove the check as well as the declaration and assignment of `tag_type` entirely.

hanno-arm replied:
@AndrzejKurek Sorry, I overlooked that we _do_ store the tag below. In that case, please disregard my comment(s).

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        MBEDTLS_ERR_ASN1_OUT_OF_DATA );
+    tag_type = **p;
+
+    if( ( ret = mbedtls_asn1_get_tag(p, end, &len_partial,
```
Minor: Space after `(`.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        return( ret );
+    }
+
+    if( **p == '\0' ) {
```
Minor: Please add a line break before `{`.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        len_partial--;
+    }
+
+    memcpy( sig->p + len_partial, *p, len_partial );
```
Blocker: This is dangerous - we cannot trust the input to be valid, and in particular we must consider the case of an ASN.1 sequence of two integers with different length. With the present code, we overflow the `sig->p` buffer if the second integer happened to be longer than the first.

AndrzejKurek replied:
Is this a valid case? I've looked into the PSA implementation, and it requires the signature to have the length of 2 * curve bytes:
`if( signature_length != 2 * curve_bytes )
        return( PSA_ERROR_INVALID_SIGNATURE );`
I can of course add checks to ensure the second length read is at most (or exactly) as big as the first.

hanno-arm replied:
It's not valid, but given that PSA is an external module I think we should err on the safe side here and check that the length values do indeed match.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    mbedtls_platform_zeroize( buf, buff_len );
+    key.pk_info = &pk_info;
+    key.pk_ctx = ctx;
+    psa_crypto_init();
```
I think we assume that the underlying PSA implementation is already initialized; it must be so for the user configuration of opaque keys, for example.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    mbedtls_pk_context key;
+    mbedtls_asn1_buf signature;
+    int key_len;
+    const int buff_len = 30 + 2 * MBEDTLS_ECP_MAX_BYTES; // Equivalent of ECP_PUB_DER_MAX_BYTES
```
We usually call this `buf_len` with only one `f`.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    int key_len;
+    const int buff_len = 30 + 2 * MBEDTLS_ECP_MAX_BYTES; // Equivalent of ECP_PUB_DER_MAX_BYTES
+    unsigned char buf[buff_len];
+    unsigned char *p = ( unsigned char* ) sig;
```
Style: Please remove the spaces after/before `(` and `)`.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    unsigned char buf[buff_len];
+    unsigned char *p = ( unsigned char* ) sig;
+    mbedtls_pk_info_t pk_info = mbedtls_eckey_info;
+    psa_algorithm_t psa_sig_md = PSA_ALG_ECDSA( translate_md_to_psa( md_alg ) );
```
We should check whether the translation succeeded.

AndrzejKurek replied:
Currently, test_suite_pk passes MBEDTLS_MD_NONE as the MD alg, and it works with PSA well, because it's getting PSA_ALG_ECDSA_BASE to its internals. What would be the better approach - letting the function work even without the md alg - as it >can< work, or require the md_alg to be set and modify the tests?

gilles-peskine-arm replied:
Relying on 0 is fragile. It just happens to work because ECDSA doesn't use a hash internally. But it doesn't work with deterministic ECDSA, for example. If you get 0 here, you should figure out why and act accordingly, not silently sign in a way that may or may not have been intended.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    unsigned char *p = ( unsigned char* ) sig;
+    mbedtls_pk_info_t pk_info = mbedtls_eckey_info;
+    psa_algorithm_t psa_sig_md = PSA_ALG_ECDSA( translate_md_to_psa( md_alg ) );
+    psa_ecc_curve_t curve = mbedtls_ecc_group_to_psa ( ( (mbedtls_ecdsa_context *) ctx )->grp.id );
```
Same here.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    mbedtls_pk_info_t pk_info = mbedtls_eckey_info;
+    psa_algorithm_t psa_sig_md = PSA_ALG_ECDSA( translate_md_to_psa( md_alg ) );
+    psa_ecc_curve_t curve = mbedtls_ecc_group_to_psa ( ( (mbedtls_ecdsa_context *) ctx )->grp.id );
+    ((void) md_alg);
```
As far as I can see, `md_alg` is always used, so this isn't needed, or is it?

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    ((void) md_alg);
+
+    memset( &signature, 0, sizeof( mbedtls_asn1_buf ) );
+    mbedtls_platform_zeroize( buf, buff_len );
```
`memset()` is sufficient here.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+         goto cleanup;
+    }
+    ret = 0;
+    psa_destroy_key( key_slot );
```
We're missing key destruction in case `psa_set_key_policy()` or `psa_import_key()` fail here. What about setting `key_slot = 0` initially and destroying it in the cleanup section as soon as it's nonzero?

AndrzejKurek replied:
Hmm, from what I understand you don't have to destroy a key if import_key will fail - could you point me to a doc stating otherwise, please? Besides, what if the returned error is PSA_ERROR_OCCUPIED_SLOT and we destroy another key (of course it shouldn't happen if `mbedtls_psa_get_free_key_slot` is implemented properly)? The PSA documentation should be more precise on which errors should result in an explicit slot destruction.
As for the set_key_policy failing, I've also assumed that the key doesn't have to be destructed (because it does not exist), but I agree that there will be the policy metadata left behind, so I'll add a psa_destroy_key for that case.

hanno-arm replied:
@AndrzejKurek I assumed that after `psa_set_key_policy()` has been successfully called, the key slot is no longer considered 'fresh', but I might be wrong. For example, it is allowed to call `psa_set_key_policy()` on the same key slot multiple times before importing key material? In regards to `psa_import_key()`, the abstract state of a key slot when `psa_import_key()` fails isn't clear to me, either.

Ping @gilles-peskine-arm.

gilles-peskine-arm replied:
Yes, you may call `psa_set_key_policy` multiple times. The slot is in a configured-but-empty state. If `psa_import_key` fails, the slot stays in this state. Call `psa_destroy_key` to wipe the configuration state.

AndrzejKurek replied:
Done.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    psa_destroy_key( key_slot );
+
+    cleanup:
+    mbedtls_free( signature.p );
```
Blocker: This can lead to double-freeing `signature.p` in case `ecdsa_extract_sig()` failed, e.g. in L.576.

AndrzejKurek replied:
Right. Since signature.p is memset to 0's, would you consider not freeing it in extract_ecdsa_sig upon error and only in the ecdsa_verify_wrap cleanup as valid?

hanno-arm replied:
@AndrzejKurek I think it's cleaner to mandate that `extract_ecdsa_sig()` passes ownership to `signature.p` only on success, guaranteeing that it is `NULL` otherwise.

AndrzejKurek replied:
Done.

mpg commented:
@hanno-arm 

> The code calls psa_init(), but I think we assume that the PSA implementation has already been initialized when Mbed TLS is loaded.

I agree, but I'm wondering it we should have some kind of documentation somewhere (maybe a KB article) about "how to use PSA with Mbed TLS" that mentions this explicitly, plus other useful facts like it's disabled by default and you need to enable it in such and such way, our general approach to minimise API changes, the new APIs available when PSA is enabled, etc.

cc: @sbutcher-arm

mpg commented:
@AndrzejKurek @hanno-arm Actually regarding `psa_init()`, I think this PR needs to add a call to it (conditional on `MBEDTLS_USE_PSA_CRYPTO`) near the beginning of the `main()` function of every single example program that might be doing any ECDSA signature. I guess that's all programs in the `pkey`, `ssl` and `x509` directories, plus `benchmark.c` and `selftest.c` in `tests`. Or it might be easier to add it to all programs just to be sure, and also so that we don't have to worry about it in future PRs.

mpg commented:
> Please note that the format of EC private keys has changed with the merge of [ARMmbed/mbedtls-psa#182](https://github.com/ARMmbed/mbedtls-psa/pull/182) . Now you need to parse the DER structure outside the PSA boundary. We intend to keep providing a function like `mbedtls_pk_parse_key` in Mbed Crypto, but how this meshes with slot allocation is still TBD.

Thanks for the info! I think however this particular PR only deals with public key. I saw in your announcement email that this is going to be changed too, is there by any chance already an open PR that we could look at (and perhaps even test our code with)?

mpg commented:
@AndrzejKurek Please rebase this over #2184 and adjust (no need to call `psa_crypto_init()` any more, and can now use some of the helper functions from `mbedtls/psa_util.h` that were so far implemented here).

mpg commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+
+    if( mbedtls_psa_get_free_key_slot( &key_slot ) != PSA_SUCCESS )
+    {
+        ret = MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE;
```
I don't think that's the most appropriate error code here. `FEATURE_UNAVAILABLE` means that the operation can never be done with that particular build of the library, while here it's more a temporary "out of resources" error.

Thinking about it, you might want to make use of [that function](https://github.com/ARMmbed/mbedtls/pull/2162/files/832f70d4714fbe281c5f16cec7b4ba430a48245a..79889bc019cc8be53f87a723f99e6df6525303cf#diff-0e47986b5bc2042422a67b92ede10ee1R764) from one of my PRs (after @hanno-arm adds it to #2184).

hanno-arm replied:
@mpg @AndrzejKurek #2184 is updated to include the PSA-to-MbedTLS error translation.

mpg commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,259 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    psa_key_policy_set_usage( &policy, PSA_KEY_USAGE_VERIFY, psa_sig_md );
+    if( psa_set_key_policy( key_slot, &policy ) != PSA_SUCCESS )
+    {
+        ret = MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE;
```
Same as above (and for other calls to PSA functions as well).

mpg commented:
@AndrzejKurek Regarding "small" r and s values, I've prepared new tests that you can find in this branch: https://github.com/mpg/mbedtls/tree/pk-ecdsa-test-short-rs Please cherry-pick the commit that adds the new tests in your branch, so that we can be confident that your code works for all these sizes of r and s.

AndrzejKurek commented:
@mpg, @hanno-arm - I have rebased the PR onto the utilities branch and addressed all of your comments. I've also incorporated Manuel's tests (thanks!). New commits start at [84befbd](https://github.com/ARMmbed/mbedtls/pull/2159/commits/84befbda6f6cc7a1a9a01517fcbb52ece182e7f2).

AndrzejKurek commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,237 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+/*
+ * Extract one signature part of an ASN.1 integer type to a given buffer
```
Hah. Had it named as "int" first, then changed it. I'll do that!

hanno-arm replied:
Also, can we `part_size` to `to_len`? Usually buffer lengths are named `xxx_len` for `xxx` the name of the buffer. 

A short description of the parameters would be helpful, too. Suggestion:
```
\p from points to the start of the buffer to read the ASN.1 encoded 
signature from, and \p is its end, used for bounds checking. 
\p to is the start of the target buffer and \p to_size is its length in bytes.
On a successful run, `*from` is increased to point past the entire ASN.1
signature.
```

AndrzejKurek replied:
I'm not sure if such an elaborate description of a static function used only in one other place is desirable. @mpg, what do you think? It's not a problem for me to add it, I just see it as an inconsistency with other static functions.

hanno-arm replied:
I think everything that's not self-explanatory should be documented, with special verbosity and consideration of style (Doxygen) for public API.

hanno-arm replied:
Sorry, my suggested documentation isn't even true. It's really just extracting the raw data from an ASN.1 encoded integer, so I think a name like `asn1_mpi_extract_raw_with_padding()` would be more descriptive.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,237 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        return( ret );
+    }
+
+    while( **from == '\0' && len_partial > 0 )
```
@mpg @AndrzejKurek Shall we check for DER-compliance here? A DER-compliant signature would never have more than one zero-byte at the beginning of the data buffer, as far as I understand.

mpg replied:
I don't think we should *check* for DER-compliance (for example, we don't need to make sure that the leading 0 byte is only present if the MSB of the next byte is set), but we can certainly *assume* it (that is, change the `while` to an `if`).

gilles-peskine-arm replied:
My hunch says no, we should be compatible with implementations that always pad `r` and `s` to their nominal length. And similarly we should also accept ASN.1 negative INTEGERs and treat the sign bit as a value bit.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,237 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    }
+
+    ( *from ) += len_partial;
+    return len_total;
```
Minor: Brackets missing.

gilles-peskine-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,237 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    mbedtls_asn1_buf signature;
+    int key_len;
+    const unsigned buf_len = 30 + 2 * MBEDTLS_ECP_MAX_BYTES; // Equivalent of ECP_PUB_DER_MAX_BYTES
+    unsigned char buf[buf_len];
```
Even when we move to C99, we'll stick to C99 without VLAs, for compatibility with compilers that implement C11 without the VLA option and with static analysis to bound the stack size.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,237 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    {
+        goto cleanup;
+    }
+    sig->len = ret;
```
If the return value for success is changed to `0`, we should increase `sig->len` by `int_size` here.

AndrzejKurek commented:
@mpg, @hanno-arm - I have addressed all of the new comments apart from the one with a static function documentation. Please re-review.

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,155 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    int ret;
+    size_t tmp_size;
+
+    if( ( end - *p ) < 1 )
```
This is not needed anymore if we don't set the tag.

AndrzejKurek replied:
Sorry, but what exactly do you mean?

hanno-arm replied:
The check was needed only to validate the subsequent tag-read from `*p`, which has been removed by now.

AndrzejKurek commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,155 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    psa_algorithm_t psa_sig_md, psa_md;
+    psa_ecc_curve_t curve = mbedtls_psa_translate_ecc_group(
+                            ( (mbedtls_ecdsa_context *) ctx )->grp.id );
+    size_t signature_part_size = ( ( (mbedtls_ecdsa_context *) ctx ) ->grp.nbits + 7 ) / 8;
```
I wasn't sure about that and, apparently, added it explicitly in one place but not another. I guess it should be removed?

hanno-arm replied:
Minor: `signature_part_size` can be made `const`

hanno-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,155 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        goto cleanup;
+    }
+
+    /* Reuse the buffer of an already imported key */
```
Very minor: The wording could be improved here, e.g. to `We don't need the exported key anymore and can reuse its buffer for signature extraction.`

mpg commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,155 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        return( MBEDTLS_ERR_ASN1_OUT_OF_DATA );
+    }
+
+    if( sig == NULL )
```
This is a static function, it doesn't need to validate its parameters, and can assume that the caller passed a valid buffer.

hanno-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -38,6 +38,38 @@ EC(DSA) verify test vector #2 (bad)
 pk_ec_test_vec:MBEDTLS_PK_ECKEY:MBEDTLS_ECP_DP_SECP192R1:"046FDD3028FA94A863CD4F78DBFF8B3AA561FC6D9CCBBCA88E0AE6FA437F5415F957542D0717FF8B84562DAE99872EF841":"546869732073686F756C64206265207468652068617368206F662061206D6573736167652E00":"30350218185B2A7FB5CD9C9A8488B119B68B47D6EC833509CE9FA1FF021900FB7D259A744A2348BD45D241A39DC915B81CC2084100FA25":MBEDTLS_ERR_ECP_VERIFY_FAILED
 
+EC(DSA) verify test vector: good, bitlen(r) = 256
+depends_on:MBEDTLS_ECP_DP_SECP256R1_ENABLED
```
Where are these test vectors taken from?

AndrzejKurek replied:
You can read about the origin of these vectors here (courtesy of @mpg):
https://github.com/ARMmbed/mbedtls/pull/2159/commits/97929534825481b93d29e884c51c8e3131588319

mpg commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,155 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+         ret = MBEDTLS_ERR_ECP_VERIFY_FAILED;
+         goto cleanup;
+    }
+    ret = 0;
```
Blocker: sorry, I should have noticed earlier, but at this point we need to verify if `p == sig + sig_len` and return either `0` or `MBEDTLS_ERR_PK_SIG_MISMATCH`.

mpg replied:
Nope, because the very reason this particular error code was introduced is so that people may choose to ignore it. As its documentation says: "The buffer **contains a valid signature** followed by more data." so we must not return it without checking the validity of the signature first - and if the signature is invalid we must return something else.

See also the documentation of `mbedtls_pk_verify()` which says the same:
```
 * \return          0 on success (signature is valid),
 *                  #MBEDTLS_ERR_PK_SIG_LEN_MISMATCH if there is a valid
 *                  signature in sig but its length is less than \p siglen,
 *                  or a specific error code.
```

gilles-peskine-arm replied:
`SIG_LEN_MISMATCH` has never worked as documented anyway. #2215

mpg commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,155 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        goto cleanup;
+    }
+
+    if( ( ret = extract_ecdsa_sig( &p, p + sig_len, buf,
```
Minor: I'd find `sig + sig_len` more readable than `p + sig_len` here: `foo + foo_len` is clearly the end of the `foo` buffer, while `p` could be pointing in the middle of the buffer according to its name (of course if we check, we'll find it currently points to the start of the buffer so it's correct, but it doesn't look obviously correct).

AndrzejKurek commented:
@mpg, @hanno-arm - updated.

mpg commented:
> Can we add a test checking that we fail on overly long signature? We could take one of the valid tests and append another byte at the end of the signature.

I was checking that but it's not going well: such a test is failing in current `development` - so return `MBEDTLS_ERR_PK_SIG_LEN_MISMATCH` was broken before this PR. I'll investigate. In the meantime, I suggest we skip testing of that part for this PR, but still try to get in right based on code review. Wdyt?

mpg commented:
@hanno-arm @AndrzejKurek Handling of trailing data after a valid signature was already broken so testing it here is out of scope. I've opened [an issue](https://github.com/ARMmbed/mbedtls/issues/2215) to track this. In the meantime, I suggest we just fix the code and leave tests for later.

AndrzejKurek commented on library/pk_wrap.c:
```diff
@@ -599,19 +591,26 @@ static int ecdsa_verify_wrap( void *ctx, mbedtls_md_type_t md_alg,
 
+    if( p != sig + sig_len )
+    {
+        ret = MBEDTLS_ERR_PK_SIG_LEN_MISMATCH;
```
Previously you've mentioned that we should return "MBEDTLS_ERR_PK_SIG_MISMATCH", but there's no such error. Which exactly did you mean?

gilles-peskine-arm replied:
`MBEDTLS_ERR_ECP_VERIFY_FAILED` in `ecdsa.c`, and there's no error for that in `pk.h`.

mpg replied:
> Edit - oh, the error is ok but the place is not. Ok.

Yes, that's what I meant, sorry for not being clear.

AndrzejKurek commented:
@mpg, @hanno-arm - fixed! Thanks again and please - re-review.

gilles-peskine-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,154 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+    mbedtls_pk_context key;
+    int key_len;
+    /* see ECP_PUB_DER_MAX_BYTES in pkwrite.c */
+    unsigned char buf[30 + 2 * MBEDTLS_ECP_MAX_BYTES];
```
At the moment, with PSA, you get the maximum size of a supported public key as `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE(PSA_VENDOR_ECC_MAX_CURVE_BITS)`. That's in the format needed for `psa_import_key`, which is currently SubjectPublicKeyInfo but will change to ECPoint soon. I'm not sure whether these two PSA size-related macros will stay in this form. The more general macro `PSA_KEY_EXPORT_MAX_SIZE` will stay but it wants a precise type, not “the largest ECC type”. As the name suggests, it's intended to size an export buffer, and in that case you can query the type with `psa_get_key_information`. Do you think `PSA_VENDOR_ECC_MAX_CURVE_BITS` should be made official? What about `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE`? At the moment they're intended as building blocks for `PSA_KEY_EXPORT_MAX_SIZE` and other macros rather than for direct use in applications.

gilles-peskine-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,154 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        goto cleanup;
+    }
+
+    if( psa_import_key( key_slot, psa_type, buf + sizeof( buf ) - key_len, key_len )
```
You should translate errors more accurately — make some inverse functions from `mbedtls_to_psa_error`. `PSA_ERROR_OUT_OF_MEMORY` shouldn't translate to `BAD_INPUT_DATA`, for example.

gilles-peskine-arm commented on library/pk_wrap.c:
```diff
@@ -472,6 +478,154 @@ static int ecdsa_can_do( mbedtls_pk_type_t type )
+        goto cleanup;
+    }
+
+    /* We don't need the exported key anymore and can
```
I suggest trying to split off the part of the function that needs to store the key and give it a local variable, and split off the part of the function that needs to store the signature and give it a local variable.
